docs: Move Nix Flakes setup documentation from README.md to the manual (#2009)
nix-community/home-manager: README.md

Nix Flakes
----------

As a NixOS module

> -----------------

Reviewer: this probably needs to be a lower-level header than Nix Flakes (same with "For unprivileged setup")

whole system (e. g. `nixos-rebuild switch --flake <path>`) instead of
using the `home-manager` command line tool.

### For unprivileged setup

If you would like to set up Home Manager as an unprivileged user:

> 1.  Place `flake.nix` under `~/.config/nixpkgs/`

Author: Thanks!

Updated. Please help me take a look if it need some rewording, as I'm not a native English speaker.

        nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
        home-manager.inputs.nixpkgs.follows = "nixpkgs";
      };
    
      outputs = { home-manager, ... }:
        let
          system = "x86_64-linux";
          username = "jdoe";
        in {
          homeConfigurations.${username} = home-manager.lib.homeManagerConfiguration {
            configuration = import ./home.nix;
            inherit system username;
            homeDirectory = "/home/${username}";
            # Update the state version as needed.
            # See the changelog here:
            # https://nix-community.github.io/home-manager/release-notes.html#sec-release-21.05
            stateVersion = "21.05";
            # # Pass extra arguments into home.nix using
            # extraSpecialArgs = {
            #   key = value;
            # };
          };
      };
    }
    ```
    *   If you would like to use the `release-21.05` branch, change the `home-manager` input url to `github:nix-community/home-manager/release-21.05` and `nixpkgs` url to `github:NixOS/nixpkgs/nixos-21.05`. Make sure to also update to the `stateVersion` option accordingly.

>     *   This example implies that the configuration is written in the file `home.nix` in the same directory level as `flake.nix`. You may change the path to `configuraion = path/to/your-home-config.nix` or specify the configuration directly.

Reviewer 2: `configuraTion`

Reviewer 2: this is basic nix stuff and even if you dont know nix, I think it's straightforward and the error message will tell you where it's looking for files. So I would remove it altogether.

Reviewer 3: > The interesting bit is that home-manager accepts a --flake argument like this:
> 
> ```
> home-manager switch --flake ~/.config/nixpkgs#jdoe
> ```

A line about the `home-manager` cli was added at the end of the Flake section of the README recently.

> There is too much text explaining how flakes work IMO. There is an official documentation for that.

I wonder whether Home Manager is the first experience people have with Nix?
I think providing a simple explanation of the flake basics is good.

Documentation about the regular workflow (updating flake.lock inputs, etc.) is also good and should be kept.

> the rest could go to the wiki and be referenced from the README. Let's try to keep in check the README length.

I agree.

whole system (e. g. `nixos-rebuild switch --flake <path>`) instead of
using the `home-manager` command line tool.

### For Unprivileged Setup

If you would like to set up home-manager as an unprivileged user:
1.  Place `flake.nix` under `~/.config/nixpkgs/`

    ```Nix
    {
      description = "Home manager configuration of jane"
    
      inputs = {
        home-manager.url = "github:nix-community/home-manager";
        # Optional, get in sync with the desired nixpkgs source
        nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
        home-manager.inputs.nixpkgs.follows = "nixpkgs";
      };
    
      outputs = { home-manager, ... }:
        let
          system = "x86_64-linux";
          username = "jdoe";
        in {

>           homeManagerConfiguration.${username} = home-manager.lib.homeManagerConfiguration {
>             configuration = import ./home.nix;
>             inherit system username;
>             homeDirectory = "/home/${username}";
>             # # Pass extra arguments into home.nix using
>             # extraSpecialArgs = {
>             #   key = value;
>             # };
>           };

Reviewer 3: ```suggestion
          homeManagerConfiguration.${username} = home-manager.lib.homeManagerConfiguration {
            configuration = import ./home.nix;
            inherit system username;
            homeDirectory = "/home/${username}";
            # Update the state version as needed.
            # See the changelog here: https://nix-community.github.io/home-manager/release-notes.html#sec-release-21.05
            stateVersion = "21.05";
            # # Pass extra arguments into home.nix using
            # extraSpecialArgs = {
            #   key = value;
            # };
          };
```

      };
    }
    ```
    *   If you would like to use the `release-20.09` branch, change the home-manager input url to `github:nix-community/home-manager/release-20.09` and nixpkgs url to `github:NixOS/nixpkgs/nixos-20.09`.

> 2.  On `~/.config/nixpkgs`, build the configuration by
> 
>     ```console
>     $ nix build .#homeManagerConfiguration.jdoe.activationPackage
>     ```
> 
>     The build result will show up in the `./result` directory
> 
> 3.  If you satisfied with the result, install and switch to the generation by
> 
>     ```console
>     $ ./result/activate
>     ```
> 
> If you are using the master branch of home-manager, switch to a new generation based on the flake by

Reviewer 3: ```suggestion
2.  Since the release `21.05`, building the configuration that resides in `~/.config/nixpkgs` is as simple as

    ```console
    $ home-manager switch --flake '~/.config/nixpkgs#jdoe'
    ```

    where `jdoe` is a configuration specified in the flake file `~/.config/nixpkgs/flake.nix`.
```

> ```console
> $ home-manager --flake-uri=[your-flake-uri] [build|switch]
> ```

Reviewer 3: ```suggestion
```

> The flake setup will not be upgraded by default when switching.
> 
> Aside from updating `flake.lock` manually by `nix-flake-update`,
> you can also pass flake-related options
> such as `--recreate-lock-file` or `--update-input [input]`
> to `home-manager` when building/switching,
> and these options will then be passed to `nix build`.

Reviewer 3: ```suggestion
The flake inputs are not upgraded automatically when switching.
The analogy to the command `home-manager --update ...` is `nix flake update`.

If updating more than one input is undesirable, the command `nix flake lock --update-input <input-name>` can be used.

You can also pass flake-related options such as `--recreate-lock-file` or `--update-input [input]`
to `home-manager` when building/switching, and these options will be forwarded to `nix build`.
```

> If you are using home-manager from other branches
> such as `release-20.09`,
> the `--flake|--flake-uri` options are currently not available.
> 
> The workaround is to use
> 
> ```console
> $ nix build .#homeManagerConfiguration.jdoe.activationPackage
> ```
> 
> to build and
> 
> ```console
> $ ./result/activate
> ```
> 
> to activate.
> 
> Upgrade by `nix flake upgrade`
> or by passing options
> such as `--recreate-lock-file` or `--update-input [input]`
> to `nix build`.
> 

Reviewer 3: ```suggestion
```

whole system (e. g. `nixos-rebuild switch --flake <path>`) instead of
using the `home-manager` command line tool.

For unprivileged setup
----------------------

> If you would like to set up Home Manager as an unprivileged user:
> 1.  Place `flake.nix` under `~/.config/nixpkgs/`

Reviewer: ```suggestion
If you would like to set up Home Manager as an unprivileged user:

1.  Place `flake.nix` under `~/.config/nixpkgs/`
```
